core/src/main/java/org/elasticsearch/common/settings/SettingsService.java

```diff
@@ -0,0 +1,201 @@
+
+    protected SettingsService(Settings settings) {
+        super(settings);
+    }
+
+    /**
+     * Applies the given settings to all listeners and rolls back the result after application. This
+     * method will not change any settings but will fail if any of the settings can't be applied.
+     */
+    public synchronized Settings dryRun(Settings settings) {
+        final Settings build = Settings.builder().put(this.settings).put(settings).build();
+        try {
+            List<RuntimeException> exceptions = new ArrayList<>();
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (RuntimeException ex) {
+                    exceptions.add(ex);
+                    logger.debug("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                }
+            }
+            // here we are exhaustive and record all settings that failed.
+            ExceptionsHelper.rethrowAndSuppress(exceptions);
+        } finally {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.rollback();
+                } catch (Exception e) {
+                    logger.warn("failed to rollback settings for [{}]", e, settingUpdater);
+                }
+            }
+        }
+        return build;
+    }
+
+    /**
+     * Applies the given settings to all the settings consumers or to none of them. The settings
+     * will be merged with the node settings before they are applied while given settings override existing node
+     * settings.
+     * @param settings the settings to apply
+     * @return the unmerged applied settings
+    */
+    public synchronized Settings applySettings(Settings settings) {
+        if (lastSettingsApplied != null && settings.equals(lastSettingsApplied)) {
+            // nothing changed in the settings, ignore
+            return settings;
+        }
+        final Settings build = Settings.builder().put(this.settings).put(settings).build();
+        boolean success = false;
+        try {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (Exception ex) {
+                    logger.warn("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                    throw ex;
+                }
+            }
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                settingUpdater.apply();
+            }
+            success = true;
+        } catch (Exception ex) {
+            logger.warn("failed to apply settings", ex);
+            throw ex;
+        } finally {
+            if (success == false) {
+                for (SettingUpdater settingUpdater : settingUpdaters) {
+                    try {
+                        settingUpdater.rollback();
```

Comment:
The more I think about rollback, the more I wonder if we could do an AtomicReference to a map with all the values so that the set is just a single thing.